### PR TITLE
Two-step reconfiguration

### PR DESCRIPTION
**Pull request by wintersteiger**
This adds two-step reconfigurations, which will allow us to easily plug in Byzantine signing.

I expect there are some things that should be moved between files/folders, just let me know if something looks like it's in the wrong place.

The whole process is triggered by `nodeInfo.status = "CatchingUp";` in `actions.js`, which I will revert back to `Trusted` once all the tests pass, so that the current behaviour doesn't change too much. Is there an easy way to make this conditional on the fact that we're running with BFT consensus?

### CODE REVIEW

File: src/consensus/aft/raft.h [hunk 1]
```diff
@@ -292,7 +316,17 @@ namespace aft
...
+      if (!catching_up && !retiring)
+      {
+        active_nodes_.insert(id());
+      }
+#ifndef NDEBUG
+      std::stringstream ss;
+      for (auto& m : active_nodes_)
+        ss << " " << m;
+      LOG_DEBUG_FMT("Active nodes:{}", ss.str());
+#endif
```
Comment (eddyashton) [inline src/consensus/aft/raft.h:329]: You can build this concatenation with `fmt::join`. Something like

```suggestion
      LOG_DEBUG_FMT("Active nodes: {}", fmt::join(active_nodes_, ", "));
```

File: src/consensus/aft/raft.h [hunk 2]
```diff
@@ -284,6 +282,32 @@ namespace aft
...
+      size_t r = (catching_up || retiring) ? 0 : 1;
+      for (const auto& [_, state] : nodes)
+      {
+        if (!state.catching_up)
+        {
+          r++;
+        }
+      }
+      return r;
+    }
```
Comment (eddyashton) [inline src/consensus/aft/raft.h:309]: Is there a reason this duplicates some of `is_eligible_voter`? While less efficient, I think this would be much more readable as

```suggestion
    size_t num_eligigble_voters() const
    {
      size_t r = 0;
      for (const auto& id, _] : nodes)
      {
        if (is_eligible_voter(id))
        {
          r++;
        }
      }
      return r;
    }
```

This could also be replaced by `return std::count_if(...)`.

File: src/consensus/aft/raft.h [hunk 3]
```diff
@@ -284,6 +282,32 @@ namespace aft
...
+    bool is_eligible_voter(const ccf::NodeId& from) const
+    {
+      if (from == state->my_node_id)
+      {
+        return !catching_up && !retiring;
+      }
+      else
+      {
+        auto it = nodes.find(from);
+        return it != nodes.end() && !it->second.catching_up;
```
Comment (eddyashton) [inline src/consensus/aft/raft.h:294]: A few questions, to help me understand the implementation:

Does the local state need to be different than the state for all the other nodes?
Do we not track whether other nodes are retiring?
Do we have an entry for `state->my_node_id` in `nodes`, and if so does `nodes[me].catching_up` always match `catching_up`?

File: src/consensus/aft/raft.h [hunk 4]
```diff
@@ -430,6 +469,11 @@ namespace aft
 
     void add_configuration(Index idx, const Configuration::Nodes& conf)
     {
+      for (auto& [id, info] : conf)
+      {
+        LOG_TRACE_FMT(
+          "CADD {} {}:{} {}", id, info.hostname, info.port, info.catching_up);
+      }
```
Comment (eddyashton) [inline src/consensus/aft/raft.h:476]: I think the logs are more readable if we're more verbose:

```suggestion
      LOG_TRACE_FMT("Adding a configuration at {} of {} nodes", idx,  conf.size());
      for (auto& [id, info] : conf)
      {
        LOG_TRACE_FMT(
          "  {} = {}:{} ({})", id, info.hostname, info.port, info.catching_up ? "passive" : "active");
      }
```

File: src/consensus/aft/raft.h [hunk 5]
```diff
@@ -2393,11 +2529,12 @@ namespace aft
 
     void become_candidate()
     {
-      if (replica_state == kv::ReplicaState::Retired)
+      if (replica_state == kv::ReplicaState::Retired || retiring)
```
Comment (eddyashton) [inline src/consensus/aft/raft.h:2532]: I'm wary of having a few bools in addition to the replica state machine. It might be the right approach, but we should confirm that.

File: src/consensus/aft/raft.h [hunk 6]
```diff
@@ -2515,19 +2676,61 @@ namespace aft
...
+      catching_up = false;
 
-      LOG_INFO_FMT(
-        "Becoming retired {}: {}", state->my_node_id, state->current_view);
+#if 0
+      // To guarantee liveness during reconfiguration, we need something of this
+      // sort (e.g. when replacing the whole network in one reconfiguration).
+      // This is not enough though; we would need to check that n/2+1 of the new
+      // nodes have seen enough of the new configuration to make sure we don't
+      // lose any transactions.
```
Comment (eddyashton) [inline src/consensus/aft/raft.h:2686]: I agree with the comment, but I don't think a local decision to delay retirement will work. My understanding is; when a node knows locally of its own removal, it should stop producing writes but otherwise continues as a normal follower/primary (it is valid and correct for a node to become leader after its removal, and share this suffix describing its own removal with the other nodes). Once this is globally committed, the node can transition to `Retired` and cease to participate (with a slight complication that if it is primary, it _should_ stick a round a little longer to try to catch up _all_ nodes, and do what it can to ensure the next primary is live in the latest configuration).

I'm not sure how this changes with catch-up. I think if a node hears global commit of its removal, it can still retire instantly - this means there are a quorum of nodes in the new configuration who can take over, so they don't need this specific node to stay active any longer.

File: src/ds/thread_messaging.h [hunk 7]
```diff
@@ -67,7 +69,14 @@ namespace threading
       ThreadMsg* current = local_msg;
       local_msg = local_msg->next;
 
-      current->cb(std::unique_ptr<ThreadMsg>(current));
+      if (current->cb)
+        current->cb(std::unique_ptr<ThreadMsg>(current));
+      else
+      {
+        current->f();
```
Comment (eddyashton) [inline src/ds/thread_messaging.h:76]: Can we give `f` a better name? It looks like its a fire-once-then-kill-me callback or something?

File: src/node/configuration_tracker.h [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+  public:
+    const NodeId& node_id;
+    ConsensusType consensus_type;
+    std::shared_ptr<kv::AbstractStore> store;
+    std::shared_ptr<enclave::RPCSessions> rpcsessions;
+    std::shared_ptr<enclave::RPCMap> rpc_map;
+    crypto::KeyPairPtr node_sign_kp;
+    const crypto::Pem& node_cert;
+    std::unordered_map<ccf::NodeId, aft::NodeState>& nodes;
+    std::list<kv::Configuration> configurations;
```
Comment (eddyashton) [inline src/node/configuration_tracker.h:65]: I don't think `node_id` and `node_cert` should be references, they should just be copied locally (they may be referring to a heap-allocated object's members right now, but they could easily come from the stack and this would break). `nodes` I don't know, that may really be referring to an map owned by someone else? But that should then be some kind of smart pointer, not a reference.

File: src/node/configuration_tracker.h [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+      rpcsessions(rpcsessions_),
+      rpc_map(rpc_map_),
+      node_sign_kp(node_sign_kp_),
+      node_cert(node_cert_),
+      nodes(nodes_)
+    {
+      if (!node_sign_kp_)
+        LOG_INFO_FMT("NO SIGNING KEY!");
+      if (!rpc_map)
+        LOG_INFO_FMT("NO RPC MAP!");
```
Comment (eddyashton) [inline src/node/configuration_tracker.h:89]: Should these be errors?

File: CMakeLists.txt [hunk 10]
```diff
@@ -679,6 +682,13 @@ if(BUILD_TESTS)
     ADDITIONAL_ARGS --ccf-version ${CCF_VERSION}
   )
 
+  add_e2e_test(
```
Comment (achamayou) [inline CMakeLists.txt:685]: Better to move that to ~803 in the loop.

File: python/ccf/tx_id.py [hunk 11]
```diff
@@ -29,3 +29,8 @@ def parse(s: Optional[str]):
         except (AttributeError, ValueError):
             pass
         return None, None
+
+    def __lt__(self, other):
```
Comment (achamayou) [inline python/ccf/tx_id.py:33]: This isn't obviously correct?

File: tests/rotation.py [hunk 12]
```diff
@@ -16,6 +16,7 @@
 @reqs.can_kill_n_nodes(1)
 def test_suspend_primary(network, args):
     primary, backup = network.find_primary_and_any_backup()
+    network.wait_for_trusted_and_committed(primary, primary)
```
Comment (achamayou) [inline tests/rotation.py:19]: Can a node be primary without being trusted and committed?

File: src/ds/thread_messaging.h [hunk 13]
```diff
@@ -14,10 +14,12 @@ namespace threading
 {
   struct ThreadMsg
   {
-    void (*cb)(std::unique_ptr<ThreadMsg>);
+    void (*cb)(std::unique_ptr<ThreadMsg>) = nullptr;
+    std::function<void()> f = nullptr;
```
Comment (ashamis) [inline src/ds/thread_messaging.h:18]: why is this specialization needed? we should try to minimize the number of ways an async function should be called

File: src/ds/thread_messaging.h [hunk 14]
```diff
@@ -276,6 +293,20 @@ namespace threading
...
+      task.add_task(new ThreadMsg(std::move(f)));
+    }
+
+    void add_task(std::function<void()>&& f)
+    {
+      add_task(
+        threading::ThreadMessaging::get_execution_thread(
+          threading::MAIN_THREAD_ID),
+        std::move(f));
+    }
```
Comment (ashamis) [inline src/ds/thread_messaging.h:308]: consider something like add_task_on_main as to me an add task without a tid implies you call it on the current thread

File: src/node/configuration_tracker.h [hunk 15]
```diff
@@ -0,0 +1,413 @@
...
+#include "node/nodes.h"
+#include "node/rpc/node_call_types.h"
+#include "node/rpc/serdes.h"
+#include "rpc/serialization.h"
+
+using namespace ccf;
+
+namespace aft
+{
+  using Index = uint64_t;
```
Comment (ashamis) [inline src/node/configuration_tracker.h:24]: This should be called seqno and can't we reuse SeqNo?

File: src/node/nodes.h [hunk 16]
```diff
@@ -17,12 +17,14 @@ namespace ccf
   enum class NodeStatus
   {
     PENDING = 0,
-    TRUSTED = 1,
-    RETIRED = 2
+    CATCHING_UP = 1,
+    TRUSTED = 2,
+    RETIRED = 3
```
Comment (ashamis) [inline src/node/nodes.h:22]: If NodeStatus is serialized anywhere and sent as a network msg or put in the ledger reordering the numbers like this will break back compat

File: src/consensus/aft/raft.h [hunk 17]
```diff
@@ -1891,6 +1978,19 @@ namespace aft
...
+        response.success == AppendEntriesResponseType::OK)
+      {
+        NodeCaughtUpMsg ncamsg = {
+          {raft_node_caught_up}, state->current_view, state->last_idx};
+        for (const auto& [node_id, _] : nodes)
+        {
+          channels->send_authenticated(
+            node_id, ccf::NodeMsgType::consensus_msg, ncamsg);
+        }
+      }
```
Comment (ashamis) [inline src/consensus/aft/raft.h:1993]: should we be only sending this if the node is catching up?

File: src/consensus/aft/raft.h [hunk 18]
```diff
@@ -2017,6 +2120,13 @@ namespace aft
         return;
       }
 
+      if (!is_eligible_voter(from))
+      {
+        LOG_INFO_FMT(
```
Comment (ashamis) [inline src/consensus/aft/raft.h:2125]: FAIL message type?

File: src/consensus/aft/raft.h [hunk 19]
```diff
@@ -2198,6 +2307,19 @@ namespace aft
         state->my_node_id.trim(),
         from.trim(),
         r.last_log_idx);
+
+      if (is_primary() && node->second.catching_up)
```
Comment (ashamis) [inline src/consensus/aft/raft.h:2311]: line 2213 already checks if the node is the leader

File: src/node/configuration_tracker.h [hunk 20]
```diff
@@ -0,0 +1,413 @@
...
+        }
+      }
+    }
+
+    static void promote_cb(
+      ConfigurationTracker& configuration_tracker,
+      NodeId from,
+      TxID txid,
+      TxID primary_txid)
+    {
```
Comment (ashamis) [inline src/node/configuration_tracker.h:402]: this is already occurring in an async callback why is another needed?

File: src/consensus/aft/raft.h [hunk 21]
```diff
@@ -2198,6 +2307,19 @@ namespace aft
...
+      {
+        threading::ThreadMessaging::thread_messaging.add_task(
+          [this, from, r]() {
```
Comment (ashamis) [inline src/consensus/aft/raft.h:2314]: why do you need to capture this? also this seems unsafe

File: src/node/configuration_tracker.h [hunk 22]
```diff
@@ -0,0 +1,413 @@
...
+          else if (nit != nodes.end() && nit->second.catching_up)
+            ss << "*";
+        }
+        ss << "]@" << c.idx << " ";
+      }
+      return ss.str();
+    }
+
+    bool record_promotion(const NodeId& node_id)
+    {
```
Comment (ashamis) [inline src/node/configuration_tracker.h:319]: Sorry i might be misunderstanding what is going on here, why does this need to go through the frontend?

File: src/node/genesis_gen.h [hunk 23]
```diff
@@ -269,8 +269,13 @@ namespace ccf
 
     void add_node(const NodeId& id, const NodeInfo& node_info)
     {
-      // Increment the node id (only used in BFT)
-      get_next_id(tx.rw(tables.values), ValueIds::NEXT_NODE_ID);
+      auto config = tx.ro(tables.config);
+      auto c0 = config->get(0);
+      if (c0 && c0->consensus == BFT)
```
Comment (ashamis) [inline src/node/genesis_gen.h:274]: Should we not throw if there is no config?

File: src/consensus/aft/test/driver.cpp [hunk 24]
```diff
@@ -98,5 +98,6 @@ int main(int argc, char** argv)
     ++lineno;
   }
 
+  threading::ThreadMessaging::thread_messaging.drop_tasks();
```
Comment (ashamis) [inline src/consensus/aft/test/driver.cpp:101]: Why is this required?

File: src/consensus/aft/async_execution.h [hunk 25]
```diff
@@ -301,4 +303,30 @@ namespace aft
     size_t size;
     OArray oarray;
   };
+
+  class NodeCatchUpCallback : public AbstractMsgCallback
+  {
+  public:
+    NodeCatchUpCallback(
```
Comment (ashamis) [inline src/consensus/aft/async_execution.h:310]: This is not following the pattern of the rest of the code. You can add a virtual method to store update_node_progress

File: src/consensus/aft/raft.h [hunk 26]
```diff
@@ -141,8 +131,10 @@ namespace aft
     size_t sig_tx_interval;
 
     // Configurations
-    std::list<Configuration> configurations;
-    std::unordered_map<ccf::NodeId, NodeState> nodes;
+    aft::ConfigurationTracker configuration_tracker;
+    std::unordered_map<ccf::NodeId, aft::NodeState> nodes;
+    bool catching_up = false;
```
Comment (ashamis) [inline src/consensus/aft/raft.h:136]: Should this be an enum? I assume you cannot be both catching up and retiring?

**Review by ashamis — APPROVED** [review]

File: include/ccf/json_handler.h [hunk 27]
```diff
@@ -261,6 +261,9 @@ namespace ccf
   inline jsonhandler::JsonAdapterResponse make_error(
     http_status status, const std::string& code, const std::string& msg)
   {
+    // Debug-only log for BFT, which may swallow frontend error results.
```
Comment (ashamis) [inline include/ccf/json_handler.h:264]: This comment does not match the implementation, maybe make keep this both CFT and BFT but i am not sure everyone would agree on this

File: src/consensus/aft/raft.h [hunk 28]
```diff
@@ -380,6 +414,11 @@ namespace aft
       return state->commit_idx;
     }
 
+    Index get_commit_idx_unsafe()
```
Comment (ashamis) [inline src/consensus/aft/raft.h:417]: Can you please change the unsafe version to just aquire a lock and then call the unsafe function

File: src/consensus/aft/raft.h [hunk 29]
```diff
@@ -1853,6 +1939,7 @@ namespace aft
 
       // After entries have been deserialised, try to commit the leader's
       // commit index and update our term history accordingly
+      LOG_DEBUG_FMT("Leader's commit index: {}", r.leader_commit_idx);
```
Comment (ashamis) [inline src/consensus/aft/raft.h:1942]: Do we need to keep this log line?

File: src/consensus/aft/raft.h [hunk 30]
```diff
@@ -2409,23 +2546,36 @@ namespace aft
 
       LOG_INFO_FMT(
         "Becoming candidate {}: {}", state->my_node_id, state->current_view);
+      LOG_DEBUG_FMT("Configurations: {}", configuration_tracker.to_string());
```
Comment (ashamis) [inline src/consensus/aft/raft.h:2549]: This might be a bit too spammy, if there is not configuration changes in progress do we really need to print this log line?

**Review by ashamis — COMMENTED** [review]
Something odd is going on with the github vscode pluggin - testing to see if this fixes it

**Review by ashamis — CHANGES_REQUESTED** [review]
sorry i did not mean to signoff - still going through the review

File: src/ds/nonstd.h [hunk 31]
```diff
@@ -7,6 +7,7 @@
 #include <cctype>
 #include <string>
 #include <string_view>
+#include <tuple>
```
Comment (ashamis) [inline src/ds/nonstd.h:10]: I might be missing something but is this required?

File: src/node/configuration_tracker.h [hunk 32]
```diff
@@ -0,0 +1,413 @@
...
+      }
+
+      LOG_TRACE_FMT("Configurations: {}", to_string());
+
+      return r;
+    }
+
+    bool rollback(Index idx)
+    {
+      LOG_TRACE_FMT("Configurations: rollback to {}", idx);
```
Comment (ashamis) [inline src/node/configuration_tracker.h:125]: Should there not be at least 1 configuration left after the rollback? (The globally committed configuration)

File: src/node/configuration_tracker.h [hunk 33]
```diff
@@ -0,0 +1,413 @@
...
+        configurations.pop_front();
+        r = true;
+      }
+
+      LOG_TRACE_FMT("Configurations: {}", to_string());
+
+      return r;
+    }
+
+    bool rollback(Index idx)
```
Comment (ashamis) [inline src/node/configuration_tracker.h:123]: This is nit: but i find this API confusing, returning false to me implies that the rollback failed not that there was nothing to rollback

File: doc/schemas/node_openapi.json [hunk 34]
```diff
@@ -277,6 +277,7 @@
       "NodeStatus": {
         "enum": [
           "Pending",
+          "CatchingUp",
```
Comment (achamayou) [inline doc/schemas/node_openapi.json:280]: I'm a bit worried about this naming, Trusted formerly implied two things:
  1. that the node was participating in replication and election quorums
  2. that it had received a copy of the ledger secrets (ie. trusted from a confidentiality perspective).

`CatchingUp` effectively corresponds to the latter but not the former, but the name doesn't convey it strongly, and the existence of a separate `Trusted` stage even suggests otherwise.

I propose:

`CatchingUp` -> `Replicating`
`Trusted` -> `Participating`

or:

`CatchingUp` -> `Learner`
`Trusted` -> `Replica`

File: include/ccf/json_handler.h (hunk 27 above)
Comment (achamayou) [reply include/ccf/json_handler.h:264]: In CFT, this results in an error returned to the client, which is why we don't typically log it. In BFT of course, most execution doesn't yield anything to the client, and so this become useful for debugging.

What happens at least once per request would typically be TRACE rather than DEBUG, but I have no objection to adding this otherwise in all cases. We know we have too much logging and we need to look into named loggers.

File: include/ccf/tx_id.h [hunk 35]
```diff
@@ -82,6 +82,11 @@ namespace ccf
 
       return tx_id;
     }
+
+    bool operator<(const TxID& other) const
+    {
+      return view < other.view || (view == other.view && seqno < other.seqno);
```
[the remark this answers is not in the document]
Comment (achamayou) [reply include/ccf/tx_id.h:88]: I am also worried about this, I don't think there is a natural order for transaction ids. The answer is probably "it depends exactly what you do with it", in which case operator overloading doesn't seem like a natural choice.

File: src/consensus/aft/raft.h (hunk 26 above)
Comment (achamayou) [reply src/consensus/aft/raft.h:136]: +1 for rolling this into the state enum unless there's an explicit reason why not.

File: src/consensus/aft/raft.h [hunk 36]
```diff
@@ -141,8 +131,10 @@ namespace aft
     size_t sig_tx_interval;
 
     // Configurations
-    std::list<Configuration> configurations;
-    std::unordered_map<ccf::NodeId, NodeState> nodes;
+    aft::ConfigurationTracker configuration_tracker;
+    std::unordered_map<ccf::NodeId, aft::NodeState> nodes;
+    bool catching_up = false;
+    bool retiring = false;
```
Comment (achamayou) [inline src/consensus/aft/raft.h:137]: I'll put up a separate PR for this, but I think a flag isn't enough and we need to keep track of more than that to ensure a single signature retirement.

File: src/consensus/aft/raft.h [hunk 37]
```diff
@@ -483,6 +525,12 @@ namespace aft
       return get_latest_configuration_unsafe().size();
     }
 
+    ccf::SeqNo get_confirmed_matching_index(const NodeId& id) const
```
Comment (achamayou) [inline src/consensus/aft/raft.h:528]: We've almost entirely moved away from magic values and towards optionals, I think we should try to stick to that.

File: include/ccf/tx_id.h (hunk 35 above)
[the remark this answers is not in the document]
Comment (wintersteiger) [reply include/ccf/tx_id.h:88]: Yeah, this is incorrect. I think it works only on the primary, but an overloaded operator is a bad choice regardless.

File: src/ds/thread_messaging.h (hunk 13 above)
Comment (wintersteiger) [reply src/ds/thread_messaging.h:18]: It isn't. I just wanted to use lambdas for the functions, but I think I ended up not using this at all.

File: src/consensus/aft/raft.h (hunk 17 above)
Comment (wintersteiger) [reply src/consensus/aft/raft.h:1993]: Good point, I think you're right, we only need this during catchup.

File: src/node/configuration_tracker.h (hunk 20 above)
Comment (wintersteiger) [reply src/node/configuration_tracker.h:402]: Another what? Am I checking some condition twice or something like that?

File: src/node/configuration_tracker.h (hunk 22 above)
Comment (wintersteiger) [reply src/node/configuration_tracker.h:319]: Recommendation by @achamayou. This makes sure the transaction is properly recorded on the ledger.

File: include/ccf/json_handler.h (hunk 27 above)
Comment (wintersteiger) [reply include/ccf/json_handler.h:264]: Yeah "for BFT" means "I'm adding this because of BFT" :-)

File: src/consensus/aft/raft.h (hunk 28 above)
Comment (wintersteiger) [reply src/consensus/aft/raft.h:417]: Not sure what you mean. The safe version gets the lock, the unsafe version is not supposed to get the lock?

File: src/consensus/aft/raft.h (hunk 29 above)
Comment (wintersteiger) [reply src/consensus/aft/raft.h:1942]: No, and there are lots of other logs that should be removed too.

File: src/consensus/aft/raft.h (hunk 30 above)
Comment (wintersteiger) [reply src/consensus/aft/raft.h:2549]: I need it, but nobody else does :-)

File: src/node/configuration_tracker.h (hunk 32 above)
Comment (wintersteiger) [reply src/node/configuration_tracker.h:125]: And there isn't? I don't think the rollback logic changed.

File: src/consensus/aft/raft.h (hunk 36 above)
Comment (wintersteiger) [reply src/consensus/aft/raft.h:137]: Yes, retirement changes aren't implemented here, because I thought we wouldn't need it. We'll need a lot more to do that properly.

File: doc/schemas/node_openapi.json (hunk 34 above)
Comment (wintersteiger) [reply doc/schemas/node_openapi.json:280]: I agree, we need to change that. To me, "Participating" sounds a bit too weak, but maybe it would be a good thing to remove the connotation with confidentiality. Maybe Replicating/Replica? Or Cloning/Replica?

File: src/node/node_to_node.h [hunk 38]
```diff
@@ -209,9 +209,16 @@ namespace ccf
       size_t size) override
     {
       auto n2n_channel = channels->get(to);
-      // Sending after a channel has been destroyed is a bug.
-      assert(n2n_channel);
-      return n2n_channel->send(type, {data, size});
+      if (!n2n_channel)
+      {
+        LOG_DEBUG_FMT(
```
Comment (jumaffre) [inline src/node/node_to_node.h:214]: Curious: why has this changed in this PR?

Comment (wintersteiger) [reply src/node/node_to_node.h:214]: That's because BFT assumes that it can reach all the nodes at all times and this was the quickest way to make it work. I'll take another look at this after the ongoing changes and if possible I'd like to make it a failure again.

File: src/node/rpc/node_frontend.h [hunk 39]
```diff
@@ -409,6 +416,93 @@ namespace ccf
...
+
+        if (!consensus)
+        {
+          return make_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            "No configured consensus");
+        }
+
+        if (
```
Comment (jumaffre) [inline src/node/rpc/node_frontend.h:441]: This should already be handled by the HTTP endpoint framework as the endpoint is a write endpoint.

File: src/consensus/aft/raft.h (hunk 3 above)
Comment (wintersteiger) [reply src/consensus/aft/raft.h:294]: Not sure, yes, and no. I had tried not to disrupt the existing code too much (e.g. `me` not in `nodes`), but after the latest discussions I'm now changing this.

File: src/ds/thread_messaging.h (hunk 7 above)
Comment (wintersteiger) [reply src/ds/thread_messaging.h:76]: `f` is the `f`unction. I just wanted to use lambdas for thread functions and add a way to make them automatically retry upon failure. I might remove all of that though.

File: src/node/rpc/node_frontend.h [hunk 40]
```diff
@@ -409,6 +416,93 @@ namespace ccf
...
+
+        auto nodes = args.tx.rw(this->network.nodes);
+
+        auto node_info = nodes->get(in.node_id);
+        if (!node_info.has_value())
+        {
+          return make_error(
+            HTTP_STATUS_INTERNAL_SERVER_ERROR,
+            ccf::errors::InternalError,
+            "Missing node state");
```
Comment (jumaffre) [inline src/node/rpc/node_frontend.h:493]:
```suggestion
            fmt::format("Unknown node {}", in.node_id));
```

File: src/node/configuration_tracker.h (hunk 8 above)
Comment (wintersteiger) [reply src/node/configuration_tracker.h:65]: The whole class is just a mental separation of the configuration-relevant parts of `Aft`. It's not helpful though, so I'll move all of this back into `Aft`.

File: src/node/configuration_tracker.h (hunk 9 above)
Comment (wintersteiger) [reply src/node/configuration_tracker.h:89]: Yes, but I'll remove this anyway.

File: tests/infra/consortium.py [hunk 41]
```diff
@@ -516,12 +518,33 @@ def check_for_service(self, remote_node, status):
                 current_status == status.value
             ), f"Service status {current_status} (expected {status.value})"
 
-    def _check_node_exists(self, remote_node, node_id, node_status=None):
-        with remote_node.client() as c:
-            r = c.get(f"/node/network/nodes/{node_id}")
+    def _check_node_exists(self, remote_node, node_id, node_status=None, retries=5):
```
Comment (jumaffre) [inline tests/infra/consortium.py:521]: The Python infra usually deals with `timeout` (in secs) rather than `retries`.

File: src/consensus/aft/raft.h [hunk 42]
```diff
@@ -2409,23 +2542,36 @@ namespace aft
...
+      LOG_DEBUG_FMT("Configurations: {}", configuration_tracker.to_string());
 
       if (consensus_type != ConsensusType::BFT)
       {
         for (auto it = nodes.begin(); it != nodes.end(); ++it)
         {
+          if (it->first == state->my_node_id)
+            continue;
+
           channels->create_channel(
```
Comment (jumaffre) [inline src/consensus/aft/raft.h:2554]: Do we need to create channels to nodes that aren't eligible?

Comment (wintersteiger) [reply src/consensus/aft/raft.h:2554]: Yes we do, during catchup we need to send them append entries and in BFT they also gossip their state to the other nodes.

File: src/node/rpc/node_frontend.h [hunk 43]
```diff
@@ -409,6 +416,93 @@ namespace ccf
...
+            fmt::format("Unknown node {}", in.node_id));
+        }
+
+        node_info->status = NodeStatus::TRUSTED;
+        nodes->put(in.node_id, *node_info);
+        return make_success(true);
+      };
+
+      make_endpoint(
+        "promote", HTTP_POST, json_adapter(promote), no_auth_required)
```
Comment (jumaffre) [inline src/node/rpc/node_frontend.h:502]:
```suggestion
        "promote", HTTP_POST, json_adapter(promote), {std::make_shared<NodeCertAuthnPolicy>()})
```

Otherwise anyone could promote a node.

Comment (wintersteiger) [reply src/node/rpc/node_frontend.h:502]: Great many thanks, I just didn't know what to use here!

File: src/consensus/aft/raft.h [hunk 44]
```diff
@@ -646,6 +697,20 @@ namespace aft
             break;
           }
 
+          case raft_node_caught_up:
+          {
+            NodeCaughtUpMsg r =
+              channels->template recv_authenticated_with_load<NodeCaughtUpMsg>(
```
Comment (jumaffre) [inline src/consensus/aft/raft.h:703]:
```suggestion
              channels->template recv_authenticated<NodeCaughtUpMsg>(
```
I think?

File: src/consensus/aft/raft.h [hunk 45]
```diff
@@ -1891,6 +1979,19 @@ namespace aft
 
       channels->send_authenticated(
         to, ccf::NodeMsgType::consensus_msg, response);
+
+      if (
+        consensus_type == ConsensusType::BFT &&
+        response.success == AppendEntriesResponseType::OK)
+      {
+        NodeCaughtUpMsg ncamsg = {
+          {raft_node_caught_up}, state->current_view, state->last_idx};
```
Comment (jumaffre) [inline src/consensus/aft/raft.h:1988]: AIUI, this new message only contains the current view and last idx, which are already sent as part of the `AppendEntriesResponse`. Couldn't we use the `AppendEntriesResponse` only for this purpose instead of creating a new message type?

File: src/consensus/aft/raft.h [hunk 46]
```diff
@@ -646,6 +697,20 @@ namespace aft
...
+          {
+            NodeCaughtUpMsg r =
+              channels->template recv_authenticated_with_load<NodeCaughtUpMsg>(
+                from, data, size);
+
+            aee = std::make_unique<NodeCatchUpCallback>(
+              from,
+              configuration_tracker,
+              TxID({r.view, r.seqno}),
+              TxID({state->current_view, state->last_idx}));
```
Comment (jumaffre) [inline src/consensus/aft/raft.h:710]: This is a pair of seqno/view but I don't think this is a valid `TxID`: `last_idx` could have been committed in the previous view but we're not at `current_view`.

File: src/node/configuration_tracker.h [hunk 47]
```diff
@@ -0,0 +1,413 @@
...
+        !(txid < primary_txid) && store != nullptr)
+      {
+        return record_promotion(id);
+      }
+
+      return true;
+    }
+
+    void update_node_progress(
+      const NodeId& from, const TxID& txid, const TxID& node_txid)
```
Comment (jumaffre) [inline src/node/configuration_tracker.h:175]: `node_txid` is unused?

File: src/consensus/aft/raft.h [hunk 48]
```diff
@@ -2198,6 +2308,19 @@ namespace aft
...
+
+      if (is_primary() && node->second.catching_up)
+      {
+        threading::ThreadMessaging::thread_messaging.add_task(
+          [this, from, r]() {
+            ConfigurationTracker::promote_cb(
+              configuration_tracker,
+              from,
+              {r.term, r.last_log_idx},
+              {state->current_view, state->commit_idx});
```
Comment (jumaffre) [inline src/consensus/aft/raft.h:2320]: Same comment here: `current_view` and `commit_idx` are not a valid `TxID`.

File: src/node/configuration_tracker.h [hunk 49]
```diff
@@ -0,0 +1,413 @@
...
+      TxID txid,
+      TxID primary_txid)
+    {
+      if (!configuration_tracker.promote_if_possible(from, txid, primary_txid))
+      {
+        threading::ThreadMessaging::thread_messaging.add_task(
+          [&configuration_tracker, from, txid, primary_txid]() {
+            ConfigurationTracker::promote_cb(
+              configuration_tracker, from, txid, primary_txid);
+          });
```
Comment (jumaffre) [inline src/node/configuration_tracker.h:409]: I'm not sure whether we want to retry forever in the unlikely event that the promotion execution fails. This will fire the next time an Append Entries Response message will be received from the peer node so maybe that's sufficient?

File: src/node/node_to_node.h (hunk 38 above)
Comment (ashamis) [reply src/node/node_to_node.h:214]: In what circumstances are we attempting to send to an unknown node?

**Comment by wintersteiger** [comment]
Replaced by a series of smaller PRs.
